### PR TITLE
Scripts/Check Mainnet Deployment Status

### PR DESCRIPTION
**Pull request by teddy-quadrata**
Add 50+ tests that:
- Check Passport State
- Check Governance State
- Check Reader State
- Check Timelock State

### CODE REVIEW

File: scripts/verify/assertMainnet.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+const AML = id("AML");
+const DID = id("DID");
+const COUNTRY = id("COUNTRY");
+const IS_BUSINESS = id("IS_BUSINESS");
+
+const TEDDY = '0xffE462ed723275eF8E7655C4883e8cD428826669';
+const DANIEL = '0x5501CC22Be0F12381489D0980f20f872e1E6bfb9';
+const TRAVIS = '0xD71bB1fF98D84ae00728f4A542Fa7A4d3257b33E';
+
+const USDC = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48';
```
Comment (fabrice-quadrata) [inline scripts/verify/assertMainnet.ts:53]: Missing DAI and USDT that we support

Comment (fabrice-quadrata) [reply scripts/verify/assertMainnet.ts:53]: can you also add WBTC to make sure taht we do NOT support it.

File: scripts/verify/assertMainnet.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    [mockIssuer] = await ethers.getSigners();
+
+    const passport = await ethers.getContractAt('QuadPassport', '0x32791980a332F1283c69660eC8e426de3aD66E7f');
+    const governance = await ethers.getContractAt('QuadGovernance', '0xA16E936425df96b9dA6125B03f19C4d34b315212')
+    const reader = await ethers.getContractAt('QuadReader', '0x7907bD4Be498cC9a7E2CF1a31dEeFCD8B132bca9')
+    const timelock = await ethers.getContractAt('IAccessControlUpgradeable', TIMELOCK);
+
+
+    // CHECK RETAIL
+    const paymentTokenAML = await reader.calculatePaymentToken(AML, USDC, TEDDY);
```
Comment (fabrice-quadrata) [inline scripts/verify/assertMainnet.ts:80]: can we add more explicit comment like:

// CHECK INDIVIDUAL QUERY PRICE

File: scripts/verify/assertMainnet.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    expect(paymentTokenDID.toString()).equals(EXPECTED_INDIVIDUAL_COST_DID.toString());
+
+    const paymentTokenIS_BUSINESS = await reader.calculatePaymentToken(IS_BUSINESS, USDC, TEDDY);
+    expect(paymentTokenIS_BUSINESS.toString()).equals(EXPECTED_INDIVIDUAL_COST_IS_BUSINESS.toString());
+
+    console.log("COMPLETE PAYMENT METHOD CHECKS");
+
+    const paymentEth = await reader.calculatePaymentETH(AML, TEDDY);
+    const resultTeddyGetAttributesETH = await reader.callStatic.getAttributesETH(TEDDY, 1, AML, {value: paymentEth});
+    expect(resultTeddyGetAttributesETH[0][0]).equals(hexZeroPad('0x01', 32)); // teddy is clean
```
Comment (fabrice-quadrata) [inline scripts/verify/assertMainnet.ts:96]: can we add all EXPECTED VALUE at the top

File: scripts/verify/assertMainnet.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    expect(springLabs.toLowerCase()).equals(EXPECTED_SPRINGLABS_ISSUER.toLowerCase());
+    expect(springLabsTreasury.toLowerCase()).equals(EXPECTED_SPRINGLABS_TREASURY.toLowerCase());
+    expect(issuerCount.toString()).equals(EXPECTED_ISSUER_COUNT);
+    expect(springLabsActivationStatus.toString()).equals(ACTIVATED);
+    expect(mintPrice.toString()).equals(EXPECTED_MINT_PRICE.toString());
+    expect(revSplitIssuer.toString()).equals(EXPECTED_REV_SPLIT_ISSUER);
+
+    console.log("COMPLETE GOV STORAGE CHECKS");
+
+    expect(await governance.hasRole(GOVERNANCE_ROLE, EXPECTED_SPRINGLABS_ISSUER)).equals(false);
```
Comment (fabrice-quadrata) [inline scripts/verify/assertMainnet.ts:127]: Let's refactor this into two loops.


- List of [ADDRESSES] who's not supposed to have [ROLE]
- List of [ADDRESSES] who's supposed to have [ROLE]

File: scripts/verify/assertMainnet.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    expect(await timelock.hasRole(DEFAULT_ADMIN_ROLE, TIMELOCK)).equals(false);
+
+    expect(await timelock.hasRole(TIMELOCK_ADMIN_ROLE, MULTISIG)).equals(false);
+    expect(await timelock.hasRole(EXECUTOR_ROLE, MULTISIG)).equals(false);
+    expect(await timelock.hasRole(PROPOSER_ROLE, MULTISIG)).equals(true);
+    expect(await timelock.hasRole(DEFAULT_ADMIN_ROLE, MULTISIG)).equals(false);
+
+    console.log("COMPLETE ACCESS ROLE CHECKS");
+
+})();
```
Comment (fabrice-quadrata) [inline scripts/verify/assertMainnet.ts:169]: What I would like to see:

**QuadPassport:**
- static test that a user could mint a Passport with an issuer sig
- static test that a user cannot mint a Passport with a non-issuer sig
- QuadReader only has READER_ROLE + No other account (deployer, timelock, QuadGovernance, QuadPassport, multisig) has READER_ROLE
- QuadGovernance only has GOVERNANCE_ROLE + No other accounts has GOVERNANCE_ROLE
- governance == QuadGovernance.address
- pendingGovernance = address(0)
- symbol = "QP";
- name = "Quadrata Passport";


QuadReader:
- governance = QuadGovernance.address
- passport = QuadPassport.address
- calculatePaymentToken returns something that seems to work

File: scripts/verify/assertMainnet.ts (hunk 1 above)
Comment (teddy-quadrata) [reply scripts/verify/assertMainnet.ts:53]: added DAI and USDT
added WBTC to ensure protocol fails when user tries to pay with it

File: scripts/verify/assertMainnet.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
+import { network } from "hardhat";
+import * as dotenv from "dotenv";
+dotenv.config();
+
+const { expect } = require("chai");
+const { constants } = require("ethers");
+const { parseEther, id, hexZeroPad, parseUnits } = require("ethers/lib/utils");
+const { ethers } = require("hardhat");
+
```
[the remark this answers is not in the document]
Comment (teddy-quadrata) [reply scripts/verify/assertMainnet.ts:9]: added:
```

const QUAD_PASSPORT = '0x32791980a332F1283c69660eC8e426de3aD66E7f';
const QUAD_GOV = '0xA16E936425df96b9dA6125B03f19C4d34b315212';
const QUAD_READER = '0x7907bD4Be498cC9a7E2CF1a31dEeFCD8B132bca9';
```

File: scripts/verify/assertMainnet.ts (hunk 2 above)
Comment (teddy-quadrata) [reply scripts/verify/assertMainnet.ts]: updated to more explicit comment

File: scripts/verify/assertMainnet.ts (hunk 3 above)
Comment (teddy-quadrata) [reply scripts/verify/assertMainnet.ts:131]: added all expectations to top of file

File: scripts/verify/assertMainnet.ts (hunk 4 above)
Comment (teddy-quadrata) [reply scripts/verify/assertMainnet.ts]: switched to more maintainable and  scaleable looping flow

File: scripts/verify/assertMainnet.ts (hunk 5 above)
Comment (teddy-quadrata) [reply scripts/verify/assertMainnet.ts:327]: done

File: scripts/verify/assertMainnet.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    const reader = await ethers.getContractAt('QuadReader', '0x7907bD4Be498cC9a7E2CF1a31dEeFCD8B132bca9')
+    const timelock = await ethers.getContractAt('IAccessControlUpgradeable', TIMELOCK);
+
+
+    // CHECK RETAIL
+    const paymentTokenAML = await reader.calculatePaymentToken(AML, USDC, TEDDY);
+    expect(paymentTokenAML.toString()).equals(EXPECTED_INDIVIDUAL_COST_AML.toString());
+
+    const paymentTokenCOUNTRY = await reader.calculatePaymentToken(COUNTRY, USDC, TEDDY);
+    expect(paymentTokenCOUNTRY.toString()).equals(EXPECTED_INDIVIDUAL_COST_COUNTRY.toString());
```
[the remark this answers is not in the document]
Comment (teddy-quadrata) [reply scripts/verify/assertMainnet.ts]: for certain stables (USDT, USDC, etc), it does return a round value. we expect this

**Review by fabrice-quadrata — APPROVED** [review]

File: .env.example [hunk 8]
```diff
@@ -2,3 +2,4 @@ ETHERSCAN_API_KEY=FILL_ME_IN
 GOERLI_URI=FILL_ME_IN
 PRIVATE_KEY=FILL_ME_IN
 MUMBAI_URI=FILL_ME_IN
+ETHEREUM_MAINNET=FILL_ME_IN
```
Comment (huy-quadrata) [inline .env.example:5]: ETHEREUM_MAINNET_URI is more descriptive

**Review by huy-quadrata — APPROVED** [review]
nit:  we could break this down into several functions with a main function driving it. this makes it easier for someone to manage since a lot of the variables could be scoped better and we can also comment out function calls as needed for testing.

lgtm though. good stuff

**Comment by teddy-quadrata** [comment]
> nit: we could break this down into several functions with a main function driving it. this makes it easier for someone to manage since a lot of the variables could be scoped better and we can also comment out function calls as needed for testing.
> 
> lgtm though. good stuff

I agree with you. There's also a lot of refactoring we could do for our test suites as well, but for now, this works and scales decently. I'm sure that in the future we'll find time to do refactoring. For now we have other priorities.

**Pull request merged**